### PR TITLE
Add suport for RISCV and the Milk-V's Pioneer

### PR DESCRIPTION
**Pull request by diehlpk**
This pull request adds support to compile Kokkos on RISC-V for the [Milk-V's Pioneer](https://www.hackster.io/news/milk-v-s-pioneer-is-a-tall-glass-of-cold-risc-v-with-64-cores-up-to-128gb-of-ram-fabf9f22747f) machine.

One can add `-DKOKKOS_ARCH_RISC64_rv64imafdcv=ON` and Kokkos builds on the machine.

I am open to discussing a better name for `RISC64_rv64imafdcv` since I am not quite happy with it.

### CODE REVIEW

**Comment by diehlpk** [comment]
refs #6323

**Review by dalg24 — CHANGES_REQUESTED** [review]
We do not take PRs against master.
Please rebase your PR onto develop and edit the destination branch.

File: cmake/kokkos_arch.cmake [hunk 1]
```diff
@@ -378,6 +378,17 @@ IF (KOKKOS_ARCH_HSW)
...
 ENDIF()
 
+IF (KOKKOS_ARCH_RISC64_rv64imafdcv)
+  COMPILER_SPECIFIC_FLAGS(
+    COMPILER_ID KOKKOS_CXX_HOST_COMPILER_ID
+        Cray    NO-VALUE-SPECIFIED
+        Intel   NO-VALUE-SPECIFIED
+        MSVC    NO-VALUE-SPECIFIED
+        NVHPC   NO-VALUE-SPECIFIED
+        DEFAULT -march=rv64imafdcv
```
Comment (dalg24) [inline cmake/kokkos_arch.cmake:388]: You require to use whitespaces instead of tabs.
See https://cloud.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-6771/1/pipeline-console/log?nodeId=25

File: cmake/kokkos_arch.cmake [hunk 2]
```diff
@@ -378,6 +378,17 @@ IF (KOKKOS_ARCH_HSW)
   )
 ENDIF()
 
+IF (KOKKOS_ARCH_RISC64_rv64imafdcv)
```
Comment (dalg24) [inline cmake/kokkos_arch.cmake:381]: Why is the suffix in lowercase?  We generally use uppercase for these.
Also you did not actually register a `KOKKOS_ARCH_OPTION()` for it.

File: core/src/impl/Kokkos_Core.cpp [hunk 3]
```diff
@@ -683,6 +683,8 @@ void pre_initialize_internal(const Kokkos::InitializationSettings& settings) {
 #elif defined(KOKKOS_ARCH_AMD_ZEN3)
   declare_configuration_metadata("architecture", "CPU architecture",
                                  "AMD_ZEN3");
+#elif defined(KOKKOS_ARCH_RISC64_rv64imafdcv)
```
Comment (dalg24) [inline core/src/impl/Kokkos_Core.cpp:686]: You would need to add a `#cmakedefine` in `cmake/KokkosCore_config.h.in` for this macro to be defined.

File: cmake/kokkos_arch.cmake (hunk 2 above)
Comment (lucbv) [inline cmake/kokkos_arch.cmake:381]: I do not like the prefix `KOKKOS_ARCH_RISC64`, I think `KOKKOS_ARCH_RISCV` is better considering that other flavors of risc exist out there that are unrelated to RISC-V.

**Review by masterleinad — COMMENTED** [review]
For this to actually be visible in `CMake` you need to add a line such as
```
DECLARE_AND_CHECK_HOST_ARCH(RISC64_rv64imafdcv               "Milk-V (RISC-V) CPUs")
```
would it make sense to just call it `MILKV` instead of `RISC64_rv64imafdcv`? How specialized is this architecture? Could it be used more broadly than in `Milk-V`?

**Comment by lucbv** [comment]
`RISC64_rv64imafdcv` is not an actual architecture, it's a set of RISC-V extensions that are required by Milk-V. I agree with @masterleinad that it is a bit too prone to user error with the given name!

File: cmake/kokkos_arch.cmake (hunk 2 above)
Comment (diehlpk) [reply cmake/kokkos_arch.cmake:381]: Ok, I can change that.

File: core/src/impl/Kokkos_Core.cpp (hunk 3 above)
Comment (diehlpk) [reply core/src/impl/Kokkos_Core.cpp:686]: Ok, I added the define

File: cmake/kokkos_arch.cmake (hunk 2 above)
Comment (diehlpk) [reply cmake/kokkos_arch.cmake:381]: No, I did not register the KOKKOS_ARCH_OPTION() yet.

**Comment by diehlpk** [comment]
> We do not take PRs against master. Please rebase your PR onto develop and edit the destination branch.

Rebasing my PR onto develop with the current master as the base has too many conflicts. I will close this pull request and do a new one.

**Comment by diehlpk** [comment]
New pull request is #6773
